### PR TITLE
Add Protect/Release tile keys

### PR DESCRIPTION
**Pull request by LiubovDidkivska**
Protect adds tile keys to protected list.
If some tile key was marked as protected, it is expected to stay in
cache, that its data can be accessible.
Release keys, which was marked before as protected. After releasing tile
keys associated data in cache could expire or be evicted.

Protect and Release functionality searches for quad tree in cache to get
data handle for each tile. When protecting tiles, at least one tile is
needed to be protected to also add quad to protected list. When
releasing all tiles should be not protected to release quad.

Added integration test with protecting and releasing tiles
keys under common quad tree.

Relates-To: OLPEDGE-2127

Signed-off-by: Liubov Didkivska <ext-liubov.didkivska@here.com>

### CODE REVIEW

File: tests/integration/olp-cpp-sdk-dataservice-read/VersionedLayerClientTest.cpp [hunk 1]
```diff
@@ -3838,6 +3838,171 @@ TEST_F(DataserviceReadVersionedLayerClientTest, PrefetchTilesAndGetAggregated) {
...
+  ON_CALL(*network_mock_, Send(IsGetRequest(URL_LOOKUP_API), _, _, _, _))
+      .WillByDefault(ReturnHttpResponse(GetResponse(http::HttpStatusCode::OK),
+                                        HTTP_RESPONSE_LOOKUP));
+  const auto protected_key = geo::TileKey::FromHereTile("23618364");
+  const auto evicted_key = geo::TileKey::FromHereTile("23618365");
+  const auto other_key = geo::TileKey::FromHereTile("23618366");
+
+  // overflow the mutable cache
+  {
+    EXPECT_CALL(*network_mock_, Send(IsGetRequest(URL_LOOKUP_API), _, _, _, _))
```
Comment (kzvieriev) [inline tests/integration/olp-cpp-sdk-dataservice-read/VersionedLayerClientTest.cpp:3923]: Here and below add SCOPED_TRACE, please.

File: olp-cpp-sdk-dataservice-read/include/olp/dataservice/read/VersionedLayerClient.h [hunk 2]
```diff
@@ -382,6 +382,32 @@ class DATASERVICE_READ_API VersionedLayerClient final {
    */
   bool IsCached(const geo::TileKey& tile) const;
 
+  /**
+   * @brief Protects tile keys from eviction.
+   *
+   * @note Only tiles which data handled are present in the cache at the time of
+   * the call can be protected
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/include/olp/dataservice/read/VersionedLayerClient.h:389]:
```suggestion
   * the call can be protected.
```

File: olp-cpp-sdk-dataservice-read/include/olp/dataservice/read/VersionedLayerClient.h [hunk 3]
```diff
@@ -382,6 +382,32 @@ class DATASERVICE_READ_API VersionedLayerClient final {
    */
   bool IsCached(const geo::TileKey& tile) const;
 
+  /**
+   * @brief Protects tile keys from eviction.
+   *
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/include/olp/dataservice/read/VersionedLayerClient.h:387]: Please add more description. What does it mean if you call this with a TileKey: will be removed from LRU, list of protected keys will be stored into cache, etc.

File: olp-cpp-sdk-dataservice-read/include/olp/dataservice/read/VersionedLayerClient.h [hunk 4]
```diff
@@ -382,6 +382,32 @@ class DATASERVICE_READ_API VersionedLayerClient final {
...
+   * @return True if some keys were successfully added to the protected list;
+   * false otherwise.
+   */
+  bool Protect(const TileKeys& tiles);
+
+  /**
+   * @brief Removes a list of tiles from protection.
+   *
+   * @note Make sure that no Protect function could be called for the same
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/include/olp/dataservice/read/VersionedLayerClient.h:401]:
```suggestion
   * @note Please make sure that `Protect` will not be called for the same
```

File: olp-cpp-sdk-dataservice-read/include/olp/dataservice/read/VersionedLayerClient.h [hunk 5]
```diff
@@ -382,6 +382,32 @@ class DATASERVICE_READ_API VersionedLayerClient final {
    */
   bool IsCached(const geo::TileKey& tile) const;
 
+  /**
+   * @brief Protects tile keys from eviction.
+   *
+   * @note Only tiles which data handled are present in the cache at the time of
+   * the call can be protected
+   *
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/include/olp/dataservice/read/VersionedLayerClient.h:390]: Please add an additional `@note` specifying that this should not be called while Release() call is in progress.

File: olp-cpp-sdk-dataservice-read/include/olp/dataservice/read/VersionedLayerClient.h [hunk 6]
```diff
@@ -382,6 +382,32 @@ class DATASERVICE_READ_API VersionedLayerClient final {
...
+   * @return True if some keys were successfully added to the protected list;
+   * false otherwise.
+   */
+  bool Protect(const TileKeys& tiles);
+
+  /**
+   * @brief Removes a list of tiles from protection.
+   *
+   * @note Make sure that no Protect function could be called for the same
+   * catalog and layer during Release.
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/include/olp/dataservice/read/VersionedLayerClient.h:402]:
```suggestion
   * catalog and layer while a `Release` call is in progress.
```

File: olp-cpp-sdk-dataservice-read/include/olp/dataservice/read/VersionedLayerClient.h [hunk 7]
```diff
@@ -382,6 +382,32 @@ class DATASERVICE_READ_API VersionedLayerClient final {
...
+   *
+   * @param tiles The list of tile keys.
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/include/olp/dataservice/read/VersionedLayerClient.h:404]:
```suggestion
   * @param tiles The list of tile keys to be released from protection.
```

File: olp-cpp-sdk-dataservice-read/include/olp/dataservice/read/VersionedLayerClient.h (hunk 2 above)
Comment (HalynaDumych) [inline olp-cpp-sdk-dataservice-read/include/olp/dataservice/read/VersionedLayerClient.h:389]:
```suggestion
   * @note You can only protect tiles which data handles are present in the cache
   * at the time of the call.
```

File: olp-cpp-sdk-dataservice-read/include/olp/dataservice/read/VersionedLayerClient.h (hunk 7 above)
Comment (HalynaDumych) [reply olp-cpp-sdk-dataservice-read/include/olp/dataservice/read/VersionedLayerClient.h:404]:
```suggestion
   * @param tiles The list of tile keys to be removed from protection.
```

File: olp-cpp-sdk-dataservice-read/include/olp/dataservice/read/VersionedLayerClient.h [hunk 8]
```diff
@@ -382,6 +382,35 @@ class DATASERVICE_READ_API VersionedLayerClient final {
...
   bool IsCached(const geo::TileKey& tile) const;
 
+  /**
+   * @brief Protects tile keys from eviction.
+   *
+   * @note You can only protect tiles which data handles are present in the
+   * cache at the time of the call.
+   *
+   * @note Please do not call `Protect` while `Release` call for the same
+   * catalog and layer is in progress.
```
Comment (HalynaDumych) [inline olp-cpp-sdk-dataservice-read/include/olp/dataservice/read/VersionedLayerClient.h:392]:
```suggestion
   * @note Please do not call `Protect` while the `Release` call for the same
   * catalog and layer is in progress.
```

File: olp-cpp-sdk-dataservice-read/include/olp/dataservice/read/VersionedLayerClient.h [hunk 9]
```diff
@@ -382,6 +382,35 @@ class DATASERVICE_READ_API VersionedLayerClient final {
...
+   * @return True if some keys were successfully added to the protected list;
+   * false otherwise.
+   */
+  bool Protect(const TileKeys& tiles);
+
+  /**
+   * @brief Removes a list of tiles from protection.
+   *
+   * @note Please make sure that `Protect` will not be called for the same
+   * catalog and layer while a `Release` call is in progress.
```
Comment (HalynaDumych) [inline olp-cpp-sdk-dataservice-read/include/olp/dataservice/read/VersionedLayerClient.h:405]:
```suggestion
   * @note Please make sure that `Protect` will not be called for the same
   * catalog and layer while the `Release` call is in progress.
```

File: olp-cpp-sdk-dataservice-read/include/olp/dataservice/read/VersionedLayerClient.h [hunk 10]
```diff
@@ -382,6 +382,35 @@ class DATASERVICE_READ_API VersionedLayerClient final {
...
+   *
+   * @param tiles The list of tile keys to be released from protection.
```
Comment (HalynaDumych) [inline olp-cpp-sdk-dataservice-read/include/olp/dataservice/read/VersionedLayerClient.h:407]:
```suggestion
   * @param tiles The list of tile keys to be removed from protection.
```

File: olp-cpp-sdk-dataservice-read/include/olp/dataservice/read/VersionedLayerClient.h [hunk 11]
```diff
@@ -382,6 +382,45 @@ class DATASERVICE_READ_API VersionedLayerClient final {
...
+   * @note You can only protect tiles which data handles are present in the
+   * cache at the time of the call.
+   *
+   * @note Please do not call `Protect` while the `Release` call for the same
+   * catalog and layer is in progress.
+   *
+   * @note Protecting tile keys mean that its data and corresponding quad tree
+   * keys will be added to protected list and stored in cache. This keys will be
+   * removed from LRU, so they could not be evicted. Also they are not expire.
+   * Quad tree stays protected if at least one tile key is protected.
```
Comment (HalynaDumych) [inline olp-cpp-sdk-dataservice-read/include/olp/dataservice/read/VersionedLayerClient.h:397]: Move to line 388 (before the note).

```suggestion
   * Protecting tile keys means that its data and corresponding quadtree
   * keys are added to the protected list and stored in the cache. These keys are
   * removed from the LRU cache, so they could not be evicted. Also, they do not expire.
   * The quadtree stays protected if at least one tile key is protected.
```

File: olp-cpp-sdk-dataservice-read/include/olp/dataservice/read/VersionedLayerClient.h [hunk 12]
```diff
@@ -382,6 +382,45 @@ class DATASERVICE_READ_API VersionedLayerClient final {
...
+  /**
+   * @brief Removes a list of tiles from protection.
+   *
+   * @note Please make sure that `Protect` will not be called for the same
+   * catalog and layer while the `Release` call is in progress.
+   *
+   * @note Releasing tile keys removes data and quad tree keys from protected
+   * list. Keys will be added to LRU, so they could be evicted. Expiration value
+   * is resored and keys could expire. Quad tree could be removed from protected
+   * list, if all tile keys are not longer protected.
```
Comment (HalynaDumych) [inline olp-cpp-sdk-dataservice-read/include/olp/dataservice/read/VersionedLayerClient.h:415]: Please move to line 409 (before the note).

```suggestion
   * Releasing tile keys removes data and quadtree keys from the protected
   * list. The keys are added to the LRU cache, so they could be evicted. Expiration value
   * is restored, and keys can expire. The quadtree can be removed from the protected
   * list if all tile keys are no longer protected.
```

**Pull request merged**